File: runtime/csharpRuntime/csharpRuntime/core/ConfigProtoSerializer.cs
using System;
using System.Reflection;
using define;

namespace ConfigProto.core
{
    public class ConfigProtoSerializer
    {
        public static T DeSerialize<T>(IJsonSerializer jsonSerializer,byte[] configContent, out string error) where T : class
        {
            error = "";
            define.ConfigTable config = null;

            try
            {
                config = jsonSerializer.DeSerialize<ConfigTable>(configContent);
            }
            catch (Exception e)
            {
                error = "error on parser config content " + e.Message;
                Logger.Instance.LogError(error);
                return null;
            }

            Type configType = ReflectionManager.Instance.GetTypeByName(config.ConfigName);
            //Type configTyep = Type.GetType(config.PackageName + "." + config.ConfigName);

            if (null == configType)
            {
                error = "error on get type by name " + config.PackageName;
                Logger.Instance.LogError(error);
                return null;
            }
            var configInstance = Activator.CreateInstance(configType);
            var pps = configType.GetFields(/*BindingFlags.Public | BindingFlags.Instance*/);
            if (pps.Length != 1)
            {
                error = "error field count " + config.PackageName;
                Logger.Instance.LogError(error);
                return null;
            }
            if (config.Type == ConfigType.TypeList)
            {
                ParserList(configInstance,config, pps[0],out error);
            }
            else
            {
                ParserMap(configInstance, config, pps[0], out error);
            }
            if (error != "")
            {
                Logger.Instance.LogError(error);
            }
            return configInstance as T;
        }

        #region parser list
        private static void ParserList(Object configInstance,ConfigTable config,FieldInfo propertyInfo, out string error)
        {
            error = "";
            string fieldName = config.ConfigName + "Info";
            if (!propertyInfo.FieldType.Name.StartsWith("List"))
            {
                error = "error field type ,type must be list, now is " + propertyInfo.FieldType.Name;
                return ;
            }
            var lineType = ReflectionManager.Instance.GetTypeByName(fieldName);
            if (null == lineType)
            {
                error = "cant' find type by name " + fieldName;
                return;
            }
            var content = Activator.CreateInstance(propertyInfo.FieldType);
            // set content line 
            propertyInfo.SetValue(configInstance, content);
            var lineMethorInfo = content.GetType().GetMethod("Add");
            if (null == lineMethorInfo)
            {
                error = "cant' find method by name Add";
                return;
            }

            string parserCellErrorMsg = "";
            for (int j = 0; j < config.Content.Count; ++j)
            {
                var line = config.Content[j];

                if (line.Content.Count != config.FieldInfoList.Count)
                {
                    Logger.Instance.LogWarning("line content count != fieldInfo count");
                    continue;
                }
                
                var lineInstance = Activator.CreateInstance(lineType);
                for (int i = 0; i < config.FieldInfoList.Count; ++i)
                {
                    string posInfo = "error at pos line: " + (j + 4) + " col: " + (i + 1)+ " ";

                    string parserErrorMsg = "";
                    var fieldInfo = config.FieldInfoList[i];
                    var contentCell = line.Content[i];
                    ParserCell(lineInstance,fieldInfo, contentCell, out parserErrorMsg);
                    if ("" != parserErrorMsg)
                    {
                        parserCellErrorMsg += posInfo + parserErrorMsg;
                    }
                }
                // do list.Add(lineInstance);
                lineMethorInfo.Invoke(content, new []{lineInstance});
            }
            error = parserCellErrorMsg;
            return ;
        }
        #endregion

        #region parser map
        private static void ParserMap(Object configInstance, ConfigTable config, FieldInfo propertyInfo, out string error)
        {
            error = "";
            
            ConfigFieldInfo keyFieldInfo = null;
            foreach (var fieldInfo in config.FieldInfoList)
            {
                if (fieldInfo.Name == config.KeyFieldName)
                {
                    keyFieldInfo = fieldInfo;
                    break;
                }
            }
            if (null == keyFieldInfo)
            {
                error = "cant' find key fieldinfo by name " + config.KeyFieldName;
                return;
            }
            string keyFieldName = ConvertFieldInfo(keyFieldInfo.Type, out error);
            if ("" != error)
            {
                return;
            }

            string fieldName = config.ConfigName + "Info";
            if (!propertyInfo.FieldType.Name.StartsWith("Dictionary"))
            {
                error = "error field type ,type must be Dictionary, now is " + propertyInfo.FieldType.Name;
                return;
            }
            var lineType = ReflectionManager.Instance.GetTypeByName(fieldName);
            if (null == lineType)
            {
                error = "cant' find type by name " + fieldName;
                return;
            }
            var content = Activator.CreateInstance(propertyInfo.FieldType);
            // set content line 
            propertyInfo.SetValue(configInstance, content);
            var ContainsKeyMethorInfo = content.GetType().GetMethod("ContainsKey");
            if (null == ContainsKeyMethorInfo)
            {
                error = "cant' find method by name ContainsKey";
                return;
            }
            var lineMethorInfo = content.GetType().GetMethod("Add");
            if (null == lineMethorInfo)
            {
                error = "cant' find method by name Add";
                return;
            }

            string parserCellErrorMsg = "";
            for (int j = 0; j < config.Content.Count; ++j)
            {
                var line = config.Content[j];

                if (line.Content.Count != config.FieldInfoList.Count)
                {
                    Logger.Instance.LogWarning("line content count != fieldInfo count");
                    continue;
                }

                var lineInstance = Activator.CreateInstance(lineType);
                Object keyInstance = null;
                for (int i = 0; i < config.FieldInfoList.Count; ++i)
                {
                    string posInfo = "error at pos line: " + (j + 4) + " col: " + (i + 1) + " ";

                    string parserErrorMsg = "";
                    var fieldInfo = config.FieldInfoList[i];
                    var contentCell = line.Content[i];
                    var cellInstance = ParserCell(lineInstance, fieldInfo, contentCell, out parserErrorMsg);
                    if ("" != parserErrorMsg)
                    {
                        parserCellErrorMsg += posInfo + parserErrorMsg;
                        continue;
                    }
                    if (fieldInfo.Name == config.KeyFieldName)
                    {
                        keyInstance = cellInstance;
                    }
                }
                if (null == keyInstance)
                {
                    parserCellErrorMsg += "error on parser may key at line " + (j + 4);
                    continue;
                }
                bool res = (bool)ContainsKeyMethorInfo.Invoke(content, new object[] { keyInstance });
                if (res)
                {
                    parserCellErrorMsg += "key already in map " + keyInstance + " at line: " + (j + 4);
                    continue;
                }

                // do map.Add(lineInstance);
                lineMethorInfo.Invoke(content, new[] { keyInstance, lineInstance });
            }
            error = parserCellErrorMsg;
            return;
        }
        #endregion

        #region parser tool
        private static Object ParserCell(Object lineInstance, ConfigFieldInfo fieldInfo, string cell, out string errorMsg)
        {
            errorMsg = "";
            var propertyInfo = lineInstance.GetType().GetField(fieldInfo.Name);
            if (null == propertyInfo)
            {
                errorMsg = "cant' find property by name " + fieldInfo.Name;
                return null;
            }
            Object cellInstance = null;
            // do parser cell
            if (fieldInfo.IsList)
            {
                var cellList = cell.Split('|');
                cellInstance = Activator.CreateInstance(propertyInfo.FieldType);
                var lineMethorInfo = cellInstance.GetType().GetMethod("Add");
                if (null == lineMethorInfo)
                {
                    errorMsg = "cant' find method by name Add at field " + fieldInfo.Name;
                    return null;
                }
                for (int i = 0; i < cellList.Length; ++i)
                {
                    var tmpCell = DoParserCell(cellList[i], fieldInfo, out errorMsg);
                    if ("" != errorMsg)
                    {
                        return null;
                    }
                    // do cellInstance.Add(tmpCell);
                    lineMethorInfo.Invoke(cellInstance, new[] { tmpCell });
                }
            }
            else
            {
                cellInstance = DoParserCell(cell, fieldInfo, out errorMsg);

            }
            propertyInfo.SetValue(lineInstance, cellInstance);
            
            return cellInstance;
        }
        private static Object DoParserCell(string cell, ConfigFieldInfo fieldInfo, out string errorMsg)
        {
            errorMsg = "";
            switch (fieldInfo.Type)
            {
                case FieldType.TypeInt32:
                    return ParserTool.Parser_Int32(cell, out errorMsg);
                case FieldType.TypeInt64:
                    return ParserTool.Parser_Int64(cell, out errorMsg);
                case FieldType.TypeFloat32:
                    return ParserTool.Parser_Float32(cell, out errorMsg);
                case FieldType.TypeFloat64:
                    return ParserTool.Parser_Float64(cell, out errorMsg);
                case FieldType.TypeBool:
                    return ParserTool.Parser_Bool(cell, out errorMsg);
                case FieldType.TypeString:
                    return ParserTool.Parser_String(cell, out errorMsg);
                case FieldType.TypeEnum:
                    return ParserTool.Parser_Int32(cell, out errorMsg);
                case FieldType.TypeDateTime:
                    return ParserTool.Parser_DateTime(cell, out errorMsg);
                case FieldType.TypeColor:
                    return ParserTool.Parser_Color(cell, out errorMsg);
                case FieldType.TypeClass:
                    break;
            }
            errorMsg = "error on parser " + cell + " unknown type " + fieldInfo.Type;
            return null;
        }
        private static string ConvertFieldInfo(FieldType fieldName,out string errorMsg)
        {
            errorMsg = "";
            switch (fieldName)
            {
                case FieldType.TypeInt32:
                    return "int";
                case FieldType.TypeInt64:
                    return "long";
                case FieldType.TypeFloat32:
                    return "float";
                case FieldType.TypeFloat64:
                    return "double";
                case FieldType.TypeBool:
                    return "bool";
                case FieldType.TypeString:
                    return "string";
                case FieldType.TypeEnum:
                    return "int";
                case FieldType.TypeDateTime:
                    return "DateTime";
                case FieldType.TypeColor:
                    return "Color";
                case FieldType.TypeClass:
                    break;
            }
            errorMsg = "unkonw field type " + fieldName;
            return "";
        }
        #endregion

    }
}

File: runtime/javaRuntime/src/main/java/core/ConfigProtoSerializer.java
package core;

import com.fasterxml.jackson.databind.ObjectMapper;
import core.define.*;

import java.lang.reflect.Field;
import java.lang.reflect.Method;

public class ConfigProtoSerializer {
    public static Object DeSerialize(byte[] content) throws Exception {
        core.define.ConfigTable configContent = null;
        Class configType = null;
        Class configLineType = null;
        Object configInstance = null;

        try {
            ObjectMapper mapper = new ObjectMapper();
            configContent = mapper.readValue(content, core.define.ConfigTable.class);

            String tmpClassName = configContent.getPackageName()+"." + configContent.getConfigName();
            configType = Class. forName(tmpClassName);
            configLineType = Class. forName(configContent.getPackageName()+"." + configContent.getConfigName()+"Info");

            configInstance = configType.newInstance();
            Field configTableField = configInstance.getClass().getDeclaredField("Content");
            Object configContentInstance = configTableField.getType().newInstance();
            configTableField.set(configInstance, configContentInstance);

            //get add method
            boolean needCheckKey = false;
            Object keyInstance = null;

            Method addMethod = null;
            Method containsKeyMethod = null;
            if(configContent.getType() == ConfigType.typeList){
                addMethod = configTableField.getType().getMethod("add", Object.class);
            }else{
                addMethod =  configTableField.getType().getMethod("put", Object.class, Object.class);
                containsKeyMethod =  configTableField.getType().getMethod("containsKey", Object.class);
                needCheckKey = true;
            }

            for(int row=0;row<configContent.getContentCount();++row){
                core.define.ConfigLine rowContent = configContent.getContent(row);
                Object lineInstance = configLineType.newInstance();

                for(int col =0 ;col <rowContent.getContentCount();++col){
                    if(col >= configContent.getFieldInfoListCount()){
                        System.out.println("col is out of range row: " + row + " col: " + col);
                        continue;
                    }
                    core.define.ConfigFieldInfo fieldElem = configContent.getFieldInfoList(col);
                    CommonError errorMsg = new CommonError();
                    errorMsg.errorMsg = "";
                    Field cellFieldElement = lineInstance.getClass().getDeclaredField(fieldElem.getName());
                    if(null == cellFieldElement){
                        System.out.println("not found field " + fieldElem.getName());
                        continue;
                    }
                    Object cellInstance = parserField(configContent,cellFieldElement.getType(),fieldElem,rowContent.getContent(col),errorMsg);
                    if(errorMsg.errorMsg != ""){
                        System.out.println(errorMsg.errorMsg + " row: " + row + " col: " + col);
                        continue;
                    }
                    cellFieldElement.set(lineInstance, cellInstance);
                    if(needCheckKey){
                        if(configContent.getKeyFieldName().equals( fieldElem.getName())){
                            keyInstance = cellInstance;
                        }
                    }
                }

                if(configContent.getType() == ConfigType.typeList){
                    addMethod.invoke(configContentInstance,lineInstance);

                }else{
                    if(null == keyInstance){
                        System.out.println("not found key of map " + configContent.getKeyFieldName() + " at row : " + row);
                        continue;
                    }
                    Boolean isContains = (Boolean)containsKeyMethod.invoke(configContentInstance,keyInstance);
                    if(isContains){
                        System.out.println("key repeated " + configContent.getKeyFieldName() + " at row : " + row + keyInstance.toString());
                        continue;
                    }
                    addMethod.invoke(configContentInstance,keyInstance,lineInstance);
                }
            }

            return configInstance;
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
            return null;
        }
    }
    private static Object parserField(core.define.ConfigTable configContent,Class fieldType,core.define.ConfigFieldInfo fieldInfo,String cell,CommonError error){
        try {
            if(fieldInfo.getIsList()){
                Object cellInstance = fieldType.newInstance();
                //get add method
                Method addMethod =fieldType.getMethod("add", Object.class);

                String[] childList = cell.split("|");
                for(int i=0;i<childList.length;++i){
                    Object cellElementInstance = doParserCell(configContent,childList[i],fieldInfo,error);
                    addMethod.invoke(cellInstance,cellElementInstance);
                }
                return cellInstance;
            }else{
                Object cellInstance = doParserCell(configContent,cell,fieldInfo,error);
                return cellInstance;
            }
        }catch (Exception e){
            error.errorMsg = "error on create field instance " + e.getMessage();
            return null;
        }
    }
    private static Object doParserCell(core.define.ConfigTable configContent,String cell, core.define.ConfigFieldInfo fieldInfo,CommonError errorMsg)
    {
        errorMsg.errorMsg = "";
        switch (fieldInfo.getType())
        {
            case typeInt32:
                return ParserTool.Parser_Int32(cell, errorMsg);
            case typeInt64:
                return ParserTool.Parser_Int64(cell,  errorMsg);
            case typeFloat32:
                return ParserTool.Parser_Float32(cell,  errorMsg);
            case typeFloat64:
                return ParserTool.Parser_Float64(cell,  errorMsg);
            case typeBool:
                return ParserTool.Parser_Bool(cell,  errorMsg);
            case typeString:
                return ParserTool.Parser_String(cell,  errorMsg);
            case typeEnum:
                String enumName = configContent.getPackageName()+"." + configContent.getConfigName()+fieldInfo.getParameter();
                return ParserTool.Parser_Enum(cell,  enumName, errorMsg);
            case typeDateTime:
                return ParserTool.Parser_DateTime(cell,  errorMsg);
            case typeColor:
                return ParserTool.Parser_Color(cell,  errorMsg);
            case typeClass:
                break;
        }
        errorMsg.errorMsg = "error on parser " + cell + " unknown type " + fieldInfo.getType();
        return null;
    }
}
